File: 백준/Silver/10974. 모든 순열/모든 순열.java
import java.io.*;
import java.util.*;


public class Main {
    static int N;
    static StringBuilder sb;
    static boolean[] visited;

    static void f(int k, String curS) {
        if (k == N) {
            sb.append(curS);
            return;
        }
        for (int i = 1; i <= N; i++) {
            if (!visited[i]) {
                visited[i] = true;
                if (k + 1 == N) {f(k + 1, curS + i+"\n");}
                else {f(k + 1, curS + i+" ");}
                visited[i] = false;
            }
        }
    }

    public static void main (String[]args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        sb = new StringBuilder();
        N = Integer.parseInt(br.readLine());
        visited = new boolean[N + 1];
//        Arrays.fill(visited, false);
        f(0, "");
        System.out.println(sb);
    }

}

File: 백준/Silver/10974. 모든 순열/모든 순열.py
def f(k,curS):
    if k==N:
        print(curS)
        return
    for i in range(1,N+1):
        if visited[i]:continue
        visited[i]=True
        if k+1==N:f(k+1,curS+str(i))
        else:f(k+1,curS+str(i)+" ")
        visited[i]=False
N = int(input())
visited = [False]*(N+1)
f(0,"")